### DISCUSSION TITLE
Using mutation effect in handler gives "type is not polymorphic enough"

### DISCUSSION THREAD

**Discussion by chtenb** [koka-lang/koka#587] [Q&A] [answered]
Consider this simple effect

```koka
pub effect sample
  ctl chance(case : a) : ()

pub fun handle-sample(action : () -> <sample> a) : <div,random> a
  var cases := Nil
  with handler
    ctl chance(case)
      cases := Cons(case, cases)
      resume(())
  action()
  choose(cases)
```

The line with the mutation statement gives the following compiler error, which I do not understand:

```
type error: type is not polymorphic enough
  context             :        handler
                        ...
                              resume(())
  term                :         chance(case)
                              cases := Cons(case, cases)
                              resume(())
  inferred type       : forall<c,e1> hnd/clause1<$b,(),sample,<local<_h>,div|e1>,c>
  is less general than: forall<c> hnd/clause1<c,(),sample,<local<_h>,div|_e>,_a>
  hint                : give a higher-rank type annotation to a function parameter?
```

How should I interpret this compiler error, and how can I implement this handler?

**Comment by TimWhiting** [comment, marked as answer]
I agree that this error message is pretty hard to read. 

It looks like you need to add the type `a` to the signature for the effect.
```
pub effect sample
  ctl chance(case : a) : ()
```
==>
```
pub effect sample<a>
  ctl chance(case : a) : ()
```

The problem with the first definition is that the type `a` could change with every call to chance, which would make `cases` a list that mixes types, which is not allowed in Koka. I guess it is still allowed in the operation definition because technically you could ignore the parameter. 
The error message is misleading because it addresses the consequences of the problem, and not the root problem itself.

We should probably disallow polymorphic variables as parameters to effect operations (or at least give a warning), though we need to allow polymorphic return values (for exceptions).
